src/cpu/dyn_cache.h

```diff
@@ -741,7 +745,12 @@ static inline void dyn_mem_execute()
 		pthread_jit_write_protect_np(true);
 #elif defined(HAVE_MPROTECT)
 	if (mprotect(cache_code_start_ptr, cache_code_size, PROT_EXEC | PROT_READ) == -1)
-		E_Exit("mprotect failed in dyn_mem_execute");		
+		E_Exit("mprotect failed in dyn_mem_execute");
+#elif defined(WIN32)
+	DWORD old = 0;
+	if (VirtualProtect(cache_code_start_ptr, cache_code_size,
+	                   PAGE_EXECUTE_READ, &old) == 0)
+		E_Exit("VirtualProtect failed in dyn_mem_execute");
```

Comment:
Is `VirtualProtect` in the hot-path of the dynrec code, and called for every cache page access/execute?

Would you say it's safe to assume that this call will almost never fail? 

If both are yes - then swapping the `if` for an `assert` (or `assertm` from support.h, which lets you still use a text message), would eliminate this branch for optimized builds but still retain this check-and-exit behavior for debug/non-release builds.  

So in the very rare case a user reports crashing on their system: we'd say: "try the debug build", which includes all the assert checks and reports where the failure's happening.

Reply:
it's called when switching between write and execute, so there is a flurry of them during a codegen, and then it's done.

Reply:
Thanks - yup, just confirmed the same with a bit of logging.  I think we should `assert` these, as their failure is going to be a very remote exception, and we will still benefit from the checks in our debug builds - if the time ever arises that these fail.

Reply:
ok, so, to confirm, something like:
```
	DWORD old_protect = 0;
	BOOL vp_res = VirtualProtect(cache_code_start_ptr, cache_code_size,
	                             PAGE_EXECUTE_READ, &old_protect);
	assert(vp_res != 0);
```

I'll do the same thing for the `mprotect` above.

Reply:
Yup - or you can wrap the call (I would assume this would work)

``` c++
assert(VirtualProtect(cache_code_start_ptr, cache_code_size,
	                             PAGE_EXECUTE_READ, &old_protect));
```

When the `assert` falls away during the optimized builds, the compiler will warn that the `vp_res` variable isn't used - so would need a MAYBE_UNUSED too:

`MAYBE_UNUSED const BOOL vp_res = ... ` 

OK with either!

Reply:
> I'll do the same thing for the mprotect above.

Nice! More de-branching :+1:

Reply:
We can't lose the `VirtualProtect` call, that's required for proper operation, like the `mprotect` for *nix.

So is this minor `if` branch optimization for the error case still useful?

Reply:
My bad! You're absolutely right :-) Variable check is the only way.

Reply:
(Oh, and until we move to C++17, each MAYBE_UNUSED generates a warning on macOS Clang warning-level=3)